flink-runtime/src/main/java/org/apache/flink/runtime/util/KeyedBudgetManager.java

```diff
@@ -0,0 +1,238 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.util;
+
+import org.apache.flink.types.Either;
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+
+/**
+ * Manages {@code long} available budget per key (allocation/release).
+ *
+ * <p>This manager gets a certain maximum {@code long} budget per key.
+ * Users can allocate some budget for some key and release it later.
+ * The manager keeps track of allocated/released budget and prevents from over-allocating.
+ *
+ * <p>There is also a paged type of allocation where a certain number of pages can be allocated from a set of keys.
+ * The page has its budget size. The manager allocates randomly from all keys of a given set.
+ * At the end, sum of pages allocated from each key is either requested number of pages or none.
+ * Only integer number of pages are allocated from each key respecting its available budget (no page spans two or more keys)
+ * or nothing is allocated reporting the maximum number of pages which could be allocated per each given key at the moment.
+ *
+ * @param <K> type of the budget key
+ */
+@ThreadSafe
+public class KeyedBudgetManager<K> {
+	private final Map<K, Long> maxBudgetByKey;
+
+	private final long defaultPageSize;
+
+	private final long totalNumberOfPages;
+
+	@GuardedBy("lock")
+	private final Map<K, Long> availableBudgetByKey;
+
+	private final Object lock = new Object();
```

Comment:
My suggestion is on `acquireBudgetForKey`, not `acquirePagedBudgetForKeys`. I could see you're considering the concurrent call on `MemoryManager.reserveMemory` based on the fact that `MemoryManager.reservedMemory` is a `ConcurrentHashMap`, and the current implementation will make such effort useless.

To be more explicit, I think the big lock is designed for "transaction" of acquiring budget from multiple keys, and we could have better handling to avoid contention for the non-transactional acquirement on single key, like acquirement for on-heap and off-heap pages could happen in parallel rather than blocking each other.

Reply:
Offline conclusion: change the implementation to `ConcurrentHashMap` as a follow up when needed.